rclrs/src/parameter.rs

```diff
@@ -3,3 +3,1285 @@ mod value;
+            value,
+            ranges,
+            map: Arc::downgrade(&builder.interface._parameter_map),
+            _marker: Default::default(),
+        })
+    }
+}
+
+/// A parameter that might not have a value, represented by `Option<T>`.
+/// This struct has ownership of the declared parameter. Additional parameter declaration will fail
+/// while this struct exists and the parameter will be undeclared when it is dropped.
+pub struct OptionalParameter<T: ParameterVariant> {
+    name: String,
+    value: Arc<RwLock<Option<ParameterValue>>>,
+    ranges: ParameterRanges,
+    map: Weak<Mutex<ParameterMap>>,
+    _marker: PhantomData<T>,
+}
+
+impl<T: ParameterVariant> Drop for OptionalParameter<T> {
+    fn drop(&mut self) {
+        // Clear the entry from the parameter map
+        if let Some(map) = self.map.upgrade() {
+            let storage = &mut map.lock().unwrap().storage;
+            storage.remove(&self.name);
+        }
+    }
+}
+
+/// A parameter that must have a value and cannot be written to
+/// This struct has ownership of the declared parameter. Additional parameter declaration will fail
+/// while this struct exists and the parameter will be undeclared when it is dropped.
+pub struct ReadOnlyParameter<T: ParameterVariant> {
+    name: String,
+    value: ParameterValue,
+    map: Weak<Mutex<ParameterMap>>,
+    _marker: PhantomData<T>,
+}
+
+impl<T: ParameterVariant> Drop for ReadOnlyParameter<T> {
+    fn drop(&mut self) {
+        // Clear the entry from the parameter map
+        if let Some(map) = self.map.upgrade() {
+            let storage = &mut map.lock().unwrap().storage;
+            storage.remove(&self.name);
+        }
+    }
+}
+
+impl<T: ParameterVariant> TryFrom<ParameterBuilder<'_, T>> for ReadOnlyParameter<T> {
+    type Error = DeclarationError;
+
+    fn try_from(builder: ParameterBuilder<T>) -> Result<Self, Self::Error> {
+        let ranges = builder.options.ranges.clone().into();
+        let default_value = builder.default_value.into();
+        ParameterInterface::validate_range_and_default_value(&ranges, Some(&default_value))?;
+        let value = builder
+            .interface
+            .get_declaration_default_value::<T>(&builder.name, &ranges, builder.tentative)?
+            .unwrap_or(default_value);
+        builder.interface.store_parameter(
+            &builder.name,
+            T::kind(),
+            DeclaredValue::ReadOnly(value.clone()),
+            builder.options.into(),
+        );
+        Ok(ReadOnlyParameter {
+            name: builder.name,
+            value,
+            map: Arc::downgrade(&builder.interface._parameter_map),
+            _marker: Default::default(),
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+struct DeclaredStorage {
+    value: DeclaredValue,
+    kind: ParameterKind,
+    options: ParameterOptionsStorage,
+}
+
+#[derive(Debug)]
+enum ParameterStorage {
+    Declared(DeclaredStorage),
+    Undeclared(Arc<RwLock<ParameterValue>>),
```

Comment:
In the current API, the only ways to access an undeclared parameter are
```
node.use_undeclared_parameters().get::<f64>("my_param");
node.use_undeclared_parameters().set("my_param", 1.23);
```
In both cases, the undeclared value will be locked and then cloned+returned (or set) with the parameter map locked the whole time. As far as I can figure we don't stand to gain anything from having an `Arc<RwLock<_>>` for the `Undeclared` variant.

It might make sense to have this `Arc` if we wanted to return some kind of `Undeclared` struct that holds onto a copy of the `Arc`, but I don't think that would be sound since an undeclared parameter could be changed to a declared parameter at any time, and holding an `Arc<RwLock<ParameterValue>>` of the old undeclared parameter would clash with that (e.g. how would the `Undeclared` struct know about the declared constraints?).

It might be possible to have an `Undeclared` struct if we wrap the entire `DeclaredStorage` in an `Arc<RwLock<_>>`, but until we want to provide such a struct I would recommend removing the `Arc<RwLock<_>>` for this.

Reply:
This is a really different usage of the words "declared" and "undeclared" than `rclcpp` uses imo. In `rclcpp`, there is no internal distinction between a declared and an undeclared parameter - some parameter either exists or it doesn't. The `allow_undeclared_parameters` flag just causes all parameter overrides to create/declare parameters automatically, instead of acting only as defaults for `declare_parameter()`. By contrast, you have a user-facing distinction between the two and do not implement `allow_undeclared_parameters`, right?

The PR in rclcpp that introduced the "declaration" concept was https://github.com/ros2/rclcpp/pull/495.

To reply to @mxgrey's [comment](https://github.com/ros2-rust/ros2_rust/pull/332#issuecomment-1751925649) on the rationale for the distinction:

> The gist is that declared parameters carry a lot more information with them. When a parameter is declared it will have some type (we say `kind` in the implementation to avoid keyword collision) and may have a range, description, or other properties. An undeclared parameter does not have any of that.
> 
> I imagine in rclcpp they probably just have "sensible default" values for all that extra information to store declared and undeclared parameters the same way. That makes sense in C++ where tagged unions are ... not pleasant (`std::variant` has terrible ergonomics). In this implementation we're taking an approach of "don't do/say more than what is explicitly needed". An undeclared parameter cannot have ranges, constraints, or descriptions because the user never provided any of those, so we should not pretend to have any of them in our implementation.

You could also have one parameter map, with parameter's fields for user-provided metadata being optionals, right?